Pegasus/sources/Geometry.cpp

```diff
@@ -212,6 +215,265 @@ bool intersection::CalculateRayAabbIntersection(double tMin, double tMax)
     return tMax > 0 && tMin < tMax;
 }
 
+bool intersection::IsTriangleContainsPoint(
+    glm::dvec3 const& triangleVertex1, glm::dvec3 const& triangleVertex2, glm::dvec3 const& triangleVertex3, glm::dvec3 const& point
+)
+{
+    double const distance = math::HyperPlane{triangleVertex1, triangleVertex2, triangleVertex3}.Distance(point);
+    if (!math::fp::IsNull(distance))
+    {
+        return false;
+    }
+
+    return IsSameSide(triangleVertex1, triangleVertex2, triangleVertex3, point)
+        && IsSameSide(triangleVertex1, triangleVertex3, triangleVertex2, point)
+        && IsSameSide(triangleVertex2, triangleVertex3, triangleVertex1, point);
+}
+
+glm::dvec3 intersection::cso::Support(Sphere const& sphere, glm::dvec3 direction)
+{
+    using namespace intersection;
+
+    Ray const ray{sphere.centerOfMass - direction * (sphere.radius + 1), direction};
+    RayIntersectionFactors intersectionFactors = CalculateRaySphereIntersectionFactors(
+        sphere.centerOfMass - ray.centerOfMass, sphere.radius, direction
+    );
+
+    glm::dvec3 const vertex = ray.centerOfMass + direction * intersectionFactors.tMax;
+    return vertex;
+}
+
+glm::dvec3 intersection::cso::Support(Box const& box, glm::dvec3 direction)
+{
+    using namespace intersection;
+
+    std::array<glm::dvec3, 8> boxVertices;
+    math::CalculateBoxVertices(box.iAxis, box.jAxis, box.kAxis, box.centerOfMass, boxVertices.begin());
+    std::sort(boxVertices.begin(), boxVertices.end(),
+        [&direction](glm::dvec3 const& a, glm::dvec3 const& b) -> bool
+    {
+        return glm::dot(a, direction) < glm::dot(b, direction);
+    });
+
+    return boxVertices.back();
```

Comment:
@ilia-glushchenko probably, but I feel like it's a tiny bit more optimal, and I don't see any reason not to do it.